pyrolite/mineral/normative.py

```diff
@@ -301,308 +302,661 @@ def CIPW_norm(df, Fe_correction=None, adjust_all=False):
+        'Il': 151.7452,
+        'Cc': 100.0892,
+        'C': 101.9613,
+        'Ru': 79.8988,
+        'Mt': 231.53860000000003,
+        'Hm': 159.6922,
+        'Mg-Di': 216.5504,
+        'Mg-Hy': 100.3887,
+        'Mg-Ol': 140.6931
+    }
+
+
+    # Convert element ppm to Oxides wt%
+
+    trace = ['F', 'Cl', 'S', 'Ni', 'Co',
+               'Sr', 'Ba', 'Rb', 'Cs', 'Li', 'Zr', 'Cr', 'V']
+
+    trace_oxides = ['CO2', 'SO3', 'F', 'Cl', 'S', 'NiO', 'CoO',
+                    'SrO', 'BaO', 'Rb2O', 'Cs2O', 'Li2O', 'ZrO2', 'Cr2O3', 'V2O3']
+
+    def calculate_oxide_1(oxide_weight, element_AW, ppm):
+        return (oxide_weight/element_AW) * ppm * (10**-4)
+
+
+    def calculate_oxide_2(oxide_weight, element_AW, ppm):
+        return (oxide_weight/(2 * element_AW)) * ppm * (10**-4)
+
+
+    for element in trace:
+        if element in ['Ni', 'Co', 'Sr', 'Ba', 'Zr']:
+            oxide_name = element_oxide[element]
+            df[oxide_name] = calculate_oxide_1(
+                                         pt.formula(oxide_name).mass,
+                                         element_AW[element],
+                                         df[element])
+        elif element in ['S', 'Cl', 'F']:
+            df[element] = df[element] * (10**-4)
+        else:
+            oxide_name = element_oxide[element]
+            df[oxide_name] = calculate_oxide_2(
+                                         pt.formula(oxide_name).mass,
+                                         element_AW[element],
+                                         df[element])
+
+
+    # Adjust majors wt% to 100% then adjust again to account for trace components
+
+    # First adjustment
+    df['intial_sum'] = df[majors].sum(axis=1)
+    adjustment_factor = 100 / df['intial_sum']
+    df[majors] = df[majors].mul(adjustment_factor, axis=0)
+
+    # Second adjustment
+    df['major_minor_sum'] = df[majors].sum(axis=1) + df[trace_oxides].sum(axis=1)
+    adjustment_factor = 100 / df['major_minor_sum']
+
+    df[majors + trace_oxides] = df[majors + trace_oxides].mul(adjustment_factor, axis=0)
+
+    # Rounding to 3 dp
+
+    df[majors + trace_oxides] = df[majors + trace_oxides].round(3)
+
+    # Calculation of some other parameters
+
+    df['Feo/MgO'] = ((2 * pt.formula("FeO").mass / pt.formula("MgO").mass) *
+                     df['Fe2O3'] + df['FeO'] / df['MgO'])
+
+    df['SI'] = 100 * df['MgO'] / (df['MgO'] + df['FeO'] + df['Fe2O3'] + df['Na2O'] + df['K2O'])
+
+    df['AR_True'] = (df['Al2O3'] + df['CaO'] + df['Na2O'] + df['K2O']) / (
+                df['Al2O3'] + df['CaO'] - df['Na2O'] - df['K2O'])
+    df['AR_False'] = (df['Al2O3'] + df['CaO'] + 2 * df['Na2O']) / (df['Al2O3'] + df['CaO'] - 2 * df['Na2O'])
+
+    df['AR'] = np.where(((df['K2O'] / df['Na2O']) >= 1) & ((df['K2O'] / df['Na2O']) <= 2.5) & (df['SiO2'] > 0.5),
+                        df['AR_True'], df['AR_False'])
+
+    df['Mg#'] = 100 * df['MgO'] / (df['MgO'] + df['FeO'])
+
+    # Mole Calculations
```

Comment:
You could use `df.pyrochem.to_molecular()` here, it should work as expected on `develop`.

Reply:
`df.pyrochem.to_molecular() ` didn't work for me - not sure I was using that function incorrectly?

Reply:
Check whether you assigned the output, it doesn't work inplace at the moment:
```python
df = df.pyrochem.to_molecular()
```

If it still doesn't work then you might be working from an earlier stage of develop, or the bug might still live. Feel free to leave this for now, and maybe just leave a comment suggesting to update later:

```python
# TODO: update to use df.pyrochem.to_molecular()
```

Reply:
added TODO